src/GHC/TypeLits/KnownNat/Solver.hs

```diff
@@ -241,19 +288,27 @@ reifyOp :: KnOp -> Type
 reifyOp (I i)          = mkNumLitTy i
 reifyOp (C (CType ty)) = ty
 reifyOp (Add x y)      = mkTyConApp typeNatAddTyCon $ reifyOp <$> [x, y]
+reifyOp (Sub x (I n)) | n < 0 = mkTyConApp typeNatAddTyCon $ reifyOp <$> [x, (I $ -n)]
+reifyOp (Sub x y)      = mkTyConApp typeNatSubTyCon $ reifyOp <$> [x, y]
 reifyOp (Mul x y)      = mkTyConApp typeNatMulTyCon $ reifyOp <$> [x, y]
 reifyOp (Exp x y)      = mkTyConApp typeNatExpTyCon $ reifyOp <$> [x, y]
 
 -- | Try to create evidence for a wanted constraint
-constraintToEvTerm :: KnownNatDefs -> [(CType,EvVar)] -> KnConstraint
+constraintToEvTerm :: KnownNatDefs -> [KnConstraint] -> [(CType,EvVar)] -> KnConstraint
                    -> Maybe (EvTerm,Ct)
-constraintToEvTerm defs kn_map (ct,cls,op) = (,ct) <$> go op
+constraintToEvTerm defs kn_givens kn_map (ct,cls,op) = (,ct) <$> go op
   where
+    isKnOp want (_,_,knOp) = want == knOp
+    knCoercion = makeKnCoercion cls
+    go want | ((ct,_,_) : _) <- filter (isKnOp want) kn_givens = pure . EvId . ctev_evar . ctEvidence $ ct
     go (I i)  = makeLitDict cls (mkNumLitTy i) i
-    go (C ty) = EvId <$> lookup ty kn_map
+    go op@(C ty) = offset go knCoercion op kn_givens $ EvId <$> lookup ty kn_map
+    go clpx | found@Just{} <- offset go knCoercion clpx kn_givens Nothing = found
     go e = do
       let (x,y,df) = case e of
             Add x' y' -> (x',y',knAddDFunId defs)
+            Sub x' (I n) | n < 0 -> (x',(I $ -n),knAddDFunId defs)
```

Comment:
Are there any tests that touch these code-paths? Are these code-paths ever supposed to be touched? or would this be a program error?